app/controllers/file_controller.rb

```diff
@@ -28,8 +41,113 @@ def download
     end
   end
 
+  # API Call to redirect to presign URL for a file.
+  # https://github.com/CDLUC3/mrt-doc/blob/master/endopoints/ui/presign-file.md
+  def presign
+    nk = storage_key_do
+    presigned = presign_get_by_node_key(nk)
+    return unless response.status == 200
+    url = presigned['url']
+    response.headers['Location'] = url
+    render status: 303, text: ''
+  end
+
+  # API to return the node id and key for a file within the storage service.
+  def storage_key
+    ret = storage_key_do
+    render status: ret['status'], json: ret.to_json
+  end
+
+  # Construct a storage key from component parts
+  def self.build_storage_key(ark, version, file)
+    "#{ark}|#{version}|#{file}"
+  end
+
+  # Encode a storage key constructed from component parts
+  def self.encode_storage_key(ark, version, file)
+    ERB::Util.url_encode(FileController.build_storage_key(ark, version, file))
+  end
+
+  def self.get_storage_presign_url(obj)
+    # Note - assume the config variable contains a slash...do not duplicate the slash
+    "#{APP_CONFIG['storage_presign_file'].chomp('/')}/#{obj[:node_id]}/#{obj[:key]}"
+  end
+
   private
 
+  # Perform database lookup for storge node and key
+  # rubocop:disable all
+  def storage_key_do
+    version = params_u(:version).to_i
+    ark = params_u(:object)
+    pathname = params_u(:file)
+
+    sql = %{
+      SELECT
```

Comment:
Its preferable to use ActiveRecord's helpers over direct SQL in the event that you ever change the underlying DB (e.g. mysql -> postgres). With that said though, I know ActiveRecord can be a massive pain when you have large queries like this.

I'd be happy to help out with that.

Reply:
Discussed in the code review and decided to defer because of the sql2 subquery.